File: src/cli/commands/member.ts
/**
 * `agt member` — Manage project members.
 */

import type { Command } from "commander";
import type { MemberRole, AgentProvider } from "../../lib/schema.js";
import { AGENT_PROVIDERS } from "../../lib/schema.js";
import { findProject } from "../../lib/project.js";
import { loadDoc, saveDoc } from "../../lib/storage.js";
import { addMember, removeMember, updateMember } from "../../lib/operations.js";
import { findMember } from "../../lib/queries.js";
import { error, success } from "../output.js";
import chalk from "chalk";

const VALID_ROLES: MemberRole[] = ["owner", "member", "agent"];

export function registerMember(program: Command): void {
  const member = program.command("member").description("Manage project members");

  member
    .command("add")
    .description("Add a member to the project")
    .argument("<name>", "Member display name")
    .option("-r, --role <role>", "Role: owner, member, or agent", "member")
    .option("-e, --email <email>", "Email address")
    .option("--provider <provider>", "Agent provider: claude-code, opencode, or custom")
    .option("--model <model>", "Agent model identifier")
    .option("--json", "Output as JSON")
    .action(
      async (
        name: string,
        opts: { role: string; email?: string; provider?: string; model?: string; json?: boolean },
      ) => {
        const paths = findProject();
        if (!paths) error("Not in an agt project. Run 'agt init' first.");

        let doc = await loadDoc(paths.dataPath);
        if (!doc) error("Project data not found.");

        if (!VALID_ROLES.includes(opts.role as MemberRole)) {
          error(`Invalid role "${opts.role}". Use: ${VALID_ROLES.join(", ")}`);
        }

        if (opts.provider && !AGENT_PROVIDERS.includes(opts.provider as AgentProvider)) {
          error(`Invalid provider "${opts.provider}". Use: ${AGENT_PROVIDERS.join(", ")}`);
        }

        // Check for duplicate name
        const existing = findMember(doc, name);
        if (existing && existing.name.toLowerCase() === name.toLowerCase()) {
          error(`A member named "${existing.name}" already exists.`);
        }

        const agentOpts =
          opts.role === "agent"
            ? { provider: opts.provider as AgentProvider | undefined, model: opts.model }
            : undefined;

        doc = addMember(
          doc,
          name,
          opts.role as MemberRole,
          opts.email ?? null,
          undefined,
          agentOpts,
        );
        await saveDoc(paths.dataPath, doc);

        const added = doc.members[doc.members.length - 1]!;

        if (opts.json) {
          const json: Record<string, unknown> = {
            id: added.id,
            name: added.name,
            role: added.role,
            email: added.email,
          };
          if (added.agentProvider) json.agentProvider = added.agentProvider;
          if (added.agentModel) json.agentModel = added.agentModel;
          console.log(JSON.stringify(json));
        } else {
          success(`Added member "${name}" (${opts.role})`);
        }
      },
    );

  member
    .command("list")
    .alias("ls")
    .description("List all project members")
    .option("--json", "Output as JSON")
    .action(async (opts: { json?: boolean }) => {
      const paths = findProject();
      if (!paths) error("Not in an agt project. Run 'agt init' first.");

      const doc = await loadDoc(paths.dataPath);
      if (!doc) error("Project data not found.");

      if (opts.json) {
        console.log(
          JSON.stringify(
            doc.members.map((m) => ({
              id: m.id,
              name: m.name,
              role: m.role,
              email: m.email,
            })),
          ),
        );
      } else {
        if (doc.members.length === 0) {
          console.log("No members.");
          return;
        }
        for (const m of doc.members) {
          const role = chalk.dim(`(${m.role})`);
          const email = m.email ? chalk.dim(` <${m.email}>`) : "";
          console.log(`  ${m.name} ${role}${email}`);
        }
      }
    });

  member
    .command("remove")
    .alias("rm")
    .description("Remove a member from the project")
    .argument("<name>", "Member name or ID")
    .option("--json", "Output as JSON")
    .action(async (nameOrId: string, opts: { json?: boolean }) => {
      const paths = findProject();
      if (!paths) error("Not in an agt project. Run 'agt init' first.");

      let doc = await loadDoc(paths.dataPath);
      if (!doc) error("Project data not found.");

      const member = findMember(doc, nameOrId);
      if (!member) error(`Member "${nameOrId}" not found.`);

      doc = removeMember(doc, member.id);
      await saveDoc(paths.dataPath, doc);

      if (opts.json) {
        console.log(JSON.stringify({ removed: member.name, id: member.id }));
      } else {
        success(`Removed member "${member.name}"`);
      }
    });

  member
    .command("update")
    .description("Update a member's role or email")
    .argument("<name>", "Member name or ID")
    .option("-r, --role <role>", "New role: owner, member, or agent")
    .option("-e, --email <email>", "New email address")
    .option("--provider <provider>", "Agent provider: claude-code, opencode, or custom")
    .option("--model <model>", "Agent model identifier")
    .option("--rename <newName>", "Rename the member")
    .option("--json", "Output as JSON")
    .action(
      async (
        nameOrId: string,
        opts: {
          role?: string;
          email?: string;
          provider?: string;
          model?: string;
          rename?: string;
          json?: boolean;
        },
      ) => {
        const paths = findProject();
        if (!paths) error("Not in an agt project. Run 'agt init' first.");

        let doc = await loadDoc(paths.dataPath);
        if (!doc) error("Project data not found.");

        const found = findMember(doc, nameOrId);
        if (!found) error(`Member "${nameOrId}" not found.`);

        if (opts.role && !VALID_ROLES.includes(opts.role as MemberRole)) {
          error(`Invalid role "${opts.role}". Use: ${VALID_ROLES.join(", ")}`);
        }

        if (opts.provider && !AGENT_PROVIDERS.includes(opts.provider as AgentProvider)) {
          error(`Invalid provider "${opts.provider}". Use: ${AGENT_PROVIDERS.join(", ")}`);
        }

        const updates: Partial<{
          name: string;
          email: string | null;
          role: MemberRole;
          agentProvider: AgentProvider;
          agentModel: string;
        }> = {};
        if (opts.role !== undefined) updates.role = opts.role as MemberRole;
        if (opts.email !== undefined) updates.email = opts.email;
        if (opts.rename !== undefined) updates.name = opts.rename;
        if (opts.provider !== undefined) updates.agentProvider = opts.provider as AgentProvider;
        if (opts.model !== undefined) updates.agentModel = opts.model;

        if (Object.keys(updates).length === 0) {
          error("No updates specified. Use --role, --email, --rename, --provider, or --model.");
        }

        doc = updateMember(doc, found.id, updates);
        await saveDoc(paths.dataPath, doc);

        if (opts.json) {
          console.log(JSON.stringify({ updated: found.name, fields: Object.keys(updates) }));
        } else {
          success(`Updated member "${found.name}"`);
        }
      },
    );
}

File: src/rust/crates/agt-cli/src/commands/member.rs
use anyhow::Result;

use agt_lib::operations;
use agt_lib::queries;
use agt_lib::schema::*;

use super::{load_project, save_project};
use crate::output;

pub fn add(
    name: String,
    role: String,
    email: Option<String>,
    provider: Option<String>,
    model: Option<String>,
    json: bool,
) -> Result<()> {
    let (paths, mut doc) = load_project()?;

    let role: MemberRole = role.parse().map_err(|e: String| anyhow::anyhow!(e))?;
    let provider: Option<AgentProvider> = provider
        .map(|p| p.parse())
        .transpose()
        .map_err(|e: String| anyhow::anyhow!(e))?;

    operations::add_member(
        &mut doc,
        &name,
        role,
        email.as_deref(),
        None,
        provider,
        model.as_deref(),
    )?;

    save_project(&paths, &mut doc)?;

    if json {
        let member = queries::find_member(&doc, &name);
        let id = member.map(|m| m.id.clone()).unwrap_or_default();
        println!(
            "{}",
            serde_json::json!({
                "ok": true,
                "name": name,
                "id": id,
            })
        );
    } else {
        println!("Added member: {} ({})", name, role.as_str());
    }
    Ok(())
}

pub fn list(json: bool) -> Result<()> {
    let (_paths, doc) = load_project()?;
    let members = queries::read_all_members(&doc);

    if json {
        println!("{}", serde_json::to_string_pretty(&members)?);
    } else {
        output::print_members_list(&members);
    }

    Ok(())
}

pub fn remove(name: String, json: bool) -> Result<()> {
    let (paths, mut doc) = load_project()?;

    let member = queries::find_member(&doc, &name)
        .ok_or_else(|| anyhow::anyhow!("Member \"{}\" not found", name))?;

    let member_name = member.name.clone();
    operations::remove_member(&mut doc, &member.id, None)?;

    save_project(&paths, &mut doc)?;

    if json {
        println!(
            "{}",
            serde_json::json!({
                "ok": true,
                "name": member_name,
            })
        );
    } else {
        println!("Removed member: {}", member_name);
    }
    Ok(())
}

pub fn update(
    name: String,
    new_name: Option<String>,
    role: Option<String>,
    email: Option<String>,
    provider: Option<String>,
    model: Option<String>,
) -> Result<()> {
    if new_name.is_none() && role.is_none() && email.is_none() && provider.is_none() && model.is_none() {
        anyhow::bail!("No updates specified. Use --name, --role, --email, --provider, or --model.");
    }

    let (paths, mut doc) = load_project()?;

    let member = queries::find_member(&doc, &name)
        .ok_or_else(|| anyhow::anyhow!("Member \"{}\" not found", name))?;

    let role: Option<MemberRole> = role
        .map(|r| r.parse())
        .transpose()
        .map_err(|e: String| anyhow::anyhow!(e))?;
    let provider: Option<AgentProvider> = provider
        .map(|p| p.parse())
        .transpose()
        .map_err(|e: String| anyhow::anyhow!(e))?;

    let member_name = member.name.clone();
    let email_update: Option<Option<&str>> = email.as_deref().map(Some);

    operations::update_member(
        &mut doc,
        &member.id,
        new_name.as_deref(),
        email_update,
        role,
        provider,
        model.as_deref(),
        None,
    )?;

    save_project(&paths, &mut doc)?;
    println!("Updated member: {}", new_name.as_deref().unwrap_or(&member_name));
    Ok(())
}
